components/feature/push/src/main/java/mozilla/components/feature/push/AutoPushFeature.kt

```diff
@@ -0,0 +1,325 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ *  License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package mozilla.components.feature.push
+
+import android.content.Context
+import android.content.SharedPreferences
+import androidx.lifecycle.LifecycleOwner
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.plus
+import kotlinx.coroutines.runBlocking
+import mozilla.appservices.push.PushError as RustPushError
+import mozilla.appservices.push.SubscriptionResponse
+import mozilla.components.concept.push.Bus
+import mozilla.components.concept.push.EncryptedPushMessage
+import mozilla.components.concept.push.PushError
+import mozilla.components.concept.push.PushProcessor
+import mozilla.components.concept.push.PushService
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.base.observer.Observable
+import mozilla.components.support.base.observer.ObserverRegistry
+import java.io.File
+import java.util.UUID
+import java.util.concurrent.Executors
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * A implementation of a [PushProcessor] that lives as a singleton. It receives messages from a service and
+ * forwards them to be decrypted and routed. Should be created with the PushProvider.
+ *
+ * <code>
+ *     class Application {
+ *          override fun onCreate() {
+ *              val feature = AutoPushFeature(context, service, configuration)
+ *              PushProvider.install(push)
+ *          }
+ *     }
+ * </code>
+ *
+ * Subscription information for each registered [PushType] can be retrieved asynchronously:
+ *
+ * <code>
+ *    feature.registerForSubscriptions(object: Bus.Observer<PushType, String> {
+ *        override fun onSubscriptionAvailable(subscription: AutoPushSubscription) { }
+ *    })
+ *    feature.subscribeForType(PushType.Services)
+ * </code>
+ *
+ * Push messages for each registered [PushType] can also be retrieved asynchronously:
```

Comment:
"can be listened for..." is probably a better way to phrase that. `retrieved` implies historical data, e.g. messages received in the past.